diff --git a/Makefile b/Makefile
index 1586436..9b2e5ee 100644
--- a/Makefile
+++ b/Makefile
@@ -664,9 +664,14 @@ ifneq (,$(findstring msvc,$(platform)))
    CFLAGS   += -MT
    CXXFLAGS += -MT
 endif
+ifeq ($(platform), emscripten)
+   CFLAGS   += -O3 -DNDEBUG
+   CXXFLAGS += -O3 -DNDEBUG
+else
    CFLAGS   += -O2 -DNDEBUG
    CXXFLAGS += -O2 -DNDEBUG
 endif
+endif
 
 LDFLAGS += $(fpic) $(SHARED)
 FLAGS += $(fpic) 
